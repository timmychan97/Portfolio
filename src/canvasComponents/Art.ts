import Canvas from "../components/Canvas";
import Scene from "./Scene";

export default class Art {
  title = "Untitled Art"
  scene?: Scene;
  sound: Record<string, string>;
  linkedCanvas?: Canvas;
  id = 0;
  isMuted = true;

  // The properties to be shown in the property panel.
  publicProperties: Record<string, string>;

  constructor(scene?: Scene) {
    this.scene = scene;
    this.sound = {};
    this.publicProperties = {};  // Empty dictionary/object literal
  }

  createScene = (ctx: CanvasRenderingContext2D) => { };

  /**
   * Plays a sound that is defined in this.sound object
   * @param soundName The name of the audio file
   * @param repeat Makes the audio loop automatically. This is the builtin function of audio tags, which might introduce gaps between loops
   * @param seamless Make the looping seamless
   */
  playSound = (soundName: string, repeat = false, seamless = false) => {
    if (this.isMuted) return;

    if (this.sound[soundName]) {
      const audio = document.getElementById("artid" + this.id + "-" + soundName) as HTMLAudioElement;
      if (audio) {
        if (!seamless) {
          audio.loop = repeat;
          audio.play();
        }
        else {
          audio.loop = repeat;
          // HTML5 audio elements is limited. There are always a small delay when running looping audio. 
          // We implement our own looping condition
          // Inspired by https://stackoverflow.com/a/22446616/7502914
          // This still has some limitations. timeupdate event not firing when touchmove event is running.
          audio.addEventListener('timeupdate', function () {
            var buffer = .44;
            if (this.currentTime > this.duration - buffer) {
              this.currentTime = 0;
              this.play();
            }
          });
          audio.play();
        }
      }
    }
  }

  replaceArtPropByStorage() {
    Object.keys(this.publicProperties).forEach(property => {
      const storedData = this.getArtPropFromSessionStorage(property);
      if (storedData) this.publicProperties[property] = storedData;
    });
  }

  getArtPropFromSessionStorage(key: string) {
    return sessionStorage.getItem(this.id + "-" + key);
  }
}