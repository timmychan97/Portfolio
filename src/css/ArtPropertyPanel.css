.ArtPropertyPanel {
    position: absolute;
    right: 0;
    bottom: 10%;
    display: flex;
    flex-direction: row;
    pointer-events: none;
    min-height: 100px;
}
.ArtPropertyPanel .property-list {
    position: relative;
    transition: 0.5s;
    background-color: var(--color-background);
    padding: 20px max(5vw, 30px) 20px calc(20px + 3vw);
    white-space: nowrap;
    box-sizing: border-box;
    pointer-events: all;
    max-width: 95vw;
}

.ArtPropertyPanel .property-list.collapse {
    transition: 0.5s;
    padding-left: 100px;
    transform: translateX(101%);
}

.ArtPropertyPanel .rightToggle {
    transition: 0.5s;
    width: max(4vw, 40px);
    z-index: 2;
    background-color: var(--color-primary-dark);
    pointer-events: all;
    display: flex;
    cursor: pointer;
}
.ArtPropertyPanel .rightToggle:hover {
    transition: 0.1s;
    background-color: var(--color-primary-hover);
}
.ArtPropertyPanel .rightToggle.collapse {
    transition: 0.5s;
    width: max(3vw, 8px);
}
.App.dark .ArtPropertyPanel .rightToggle{
    transition: 0.1s;
    background-color: var(--color-secondary-dark);
}
.App.dark .ArtPropertyPanel .rightToggle:hover{
    transition: 0.1s;
    background-color: var(--color-secondary-hover);
}
.ArtPropertyPanel .rightToggle .arrow-icon {
    transition: 0.5s;
    position: relative;
    transform: scaleX(-1);
}
.ArtPropertyPanel .rightToggle .arrow-icon span {
    transition: 0.5s;
    background-color: var(--color-background);
}
.App.dark .ArtPropertyPanel .rightToggle .arrow-icon span {
    transition: 0.5s;
    background-color: var(--color-primary-dark);
}
.ArtPropertyPanel .rightToggle.collapse .arrow-icon span {
    transition: 0.5s;
    background-color: #00000000;
}
.ArtPropertyPanel .rightToggle .icon-wrapper {
    display: flex;
    height: 100%;
    align-items: center;
    overflow-x: visible;
}



.property-list .leftToggle {
    transition: 0.1s;
    position: absolute;
    top:0;
    left:0;
    width: 30px;
    height: 100%;
    z-index: 1;
    display: flex;
    align-items: center;
    border-right: 2px dashed  var(--color-primary-dark);
    cursor: pointer;
}

.property-list .leftToggle:hover {
    transition: 0.1s;
    background-color: #88888833;
}

.property-list .leftToggle.collapse {
    transition: 0.5s;
    width: max(3vw, 8px);
}

.App.dark .ArtPropertyPanel .leftToggle {
    transition: 0.1s;
    background-color: var(--color-secondary-dark);
    border-right: none;
}

.App.dark .ArtPropertyPanel .leftToggle:hover {
    transition: 0.1s;
    background-color: var(--color-secondary-hover);
}

.arrow-icon {
    width: 30px;
    height: 30px;
}

.arrow-icon span {
    display: block;
    position: absolute;
    height: 3px;
    width: 100%;
    background-color: var(--color-primary-dark);
    border-radius: 9px;
    opacity: 1;
    left: 0;
    transform: rotate(0deg);
}

.arrow-icon span:nth-child(1) {
    margin-bottom: 27px;
    transform: rotate(70deg);
}

.arrow-icon span:nth-child(2) {
    margin-top: 27px;
    transform: rotate(-70deg);
}




.ArtPropertyPanel ul {
    padding: 0;
    padding-left: 10px;
    margin: 4px 0px;
}
.ArtPropertyPanel li {
    display: flex;
}

/* Add gap between li elements */
.ArtPropertyPanel li + li {
    margin-top: 10px;
}

.ArtPropertyPanel li label {
    flex-grow: 1;
    padding-right: 12px;
    text-align: left;
}

.ArtPropertyPanel li input {
    width: 100px;
    max-width: 30vw;
}