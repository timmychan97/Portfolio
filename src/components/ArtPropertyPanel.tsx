import React from 'react';
import Art from '../canvasComponents/Art';
import "../css/ArtPropertyPanel.css";

interface Props {
  art: Art,
}

interface States {
  collapsed: boolean,
}

export default class ArtPropertyPanel extends React.Component<Props, States> {
  state = {
    collapsed: false,
  }

  togglePanel() {
    this.setState({ collapsed: !this.state.collapsed });
  }

  onChangeProperty(property: string, value: string) {
    this.props.art.publicProperties[property] = value;
    sessionStorage.setItem(this.props.art.id + "-" + property, value);
  }


  render() {
    return (
      <div className="ArtPropertyPanel" onClick={(e) => e.stopPropagation()}>
        <div className={"property-list" + (this.state.collapsed ? " collapse" : "")} >
          <div className={"leftToggle"} onClick={(e) => this.togglePanel()}>
            <div className="arrow-icon">
              <span /><span />
            </div>
          </div>
          <ul>
            {Object.keys(this.props.art.publicProperties).map((property, i) => {
              return (
                <li key={i}>
                  <label htmlFor={property}>{property + ":"}</label>
                  <input name={property} type="text"
                    defaultValue={this.props.art.publicProperties[property]}
                    onChange={(e) => this.onChangeProperty(property, e.target.value)
                    } />
                </li>
              );
            })}
          </ul>
        </div>
        <div className={"rightToggle" + (this.state.collapsed ? "" : " collapse")} onClick={(e) => this.togglePanel()}>
          <div className="icon-wrapper">
            <div className="arrow-icon">
              <span /><span />
            </div>
          </div>
        </div>
      </div>
    );
  }
}
