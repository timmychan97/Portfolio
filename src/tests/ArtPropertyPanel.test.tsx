import { render } from "@testing-library/react";
import React from "react";
import Art from "../canvasComponents/Art";
import ArtPropertyPanel from "../components/ArtPropertyPanel";

test('Properties matches snapshot', () => {
  const testArt = new Art();
  testArt.id = -100;
  testArt.title = "Test 1";
  testArt.publicProperties["X"] = "200";
  testArt.publicProperties["Y"] = "620";
  testArt.publicProperties["Hello world"] = "world hello";
  testArt.publicProperties["I am a test"] = "7812&T%/fds1  1";

  const { container } = render(<ArtPropertyPanel art={testArt} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('Properties matches snapshot', () => {
  const testArt = new Art();
  testArt.id = 12410;
  testArt.title = "Test 2";
  testArt.publicProperties["Xf2"] = "200";
  testArt.publicProperties["ag23Y"] = "62fs0";
  testArt.publicProperties["Hellogrg world"] = "worl1d hello   ";
  testArt.publicProperties["%&/fd"] = "Â£$24279yhnuf";

  const { container } = render(<ArtPropertyPanel art={testArt} />);
  expect(container.firstChild).toMatchSnapshot();
});