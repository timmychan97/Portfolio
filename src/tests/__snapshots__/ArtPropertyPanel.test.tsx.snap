// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properties matches snapshot 1`] = `
<div
  class="ArtPropertyPanel"
>
  <div
    class="property-list"
  >
    <div
      class="leftToggle"
    >
      <div
        class="arrow-icon"
      >
        <span />
        <span />
      </div>
    </div>
    <ul>
      <li>
        <label
          for="X"
        >
          X:
        </label>
        <input
          name="X"
          type="text"
          value="200"
        />
      </li>
      <li>
        <label
          for="Y"
        >
          Y:
        </label>
        <input
          name="Y"
          type="text"
          value="620"
        />
      </li>
      <li>
        <label
          for="Hello world"
        >
          Hello world:
        </label>
        <input
          name="Hello world"
          type="text"
          value="world hello"
        />
      </li>
      <li>
        <label
          for="I am a test"
        >
          I am a test:
        </label>
        <input
          name="I am a test"
          type="text"
          value="7812&T%/fds1  1"
        />
      </li>
    </ul>
  </div>
  <div
    class="rightToggle collapse"
  >
    <div
      class="icon-wrapper"
    >
      <div
        class="arrow-icon"
      >
        <span />
        <span />
      </div>
    </div>
  </div>
</div>
`;

exports[`Properties matches snapshot 2`] = `
<div
  class="ArtPropertyPanel"
>
  <div
    class="property-list"
  >
    <div
      class="leftToggle"
    >
      <div
        class="arrow-icon"
      >
        <span />
        <span />
      </div>
    </div>
    <ul>
      <li>
        <label
          for="Xf2"
        >
          Xf2:
        </label>
        <input
          name="Xf2"
          type="text"
          value="200"
        />
      </li>
      <li>
        <label
          for="ag23Y"
        >
          ag23Y:
        </label>
        <input
          name="ag23Y"
          type="text"
          value="62fs0"
        />
      </li>
      <li>
        <label
          for="Hellogrg world"
        >
          Hellogrg world:
        </label>
        <input
          name="Hellogrg world"
          type="text"
          value="worl1d hello   "
        />
      </li>
      <li>
        <label
          for="%&/fd"
        >
          %&/fd:
        </label>
        <input
          name="%&/fd"
          type="text"
          value="Â£$24279yhnuf"
        />
      </li>
    </ul>
  </div>
  <div
    class="rightToggle collapse"
  >
    <div
      class="icon-wrapper"
    >
      <div
        class="arrow-icon"
      >
        <span />
        <span />
      </div>
    </div>
  </div>
</div>
`;
